File: ios/Paperback/UI/ReadabilityStyle.swift
import SwiftUI

/// Applies the reader's readability preferences to document text: size, line spacing, the
/// spacing between paragraphs, and alignment.
///
/// The size is a multiplier on the Dynamic Type body size rather than a fixed point size, so a
/// reader who has already enlarged text system-wide keeps that and scales it further here.
struct ReadabilityStyle: ViewModifier {
	let scalePercent: Int
	let lineSpacingChoice: Int
	let paragraphSpacingChoice: Int
	let alignmentChoice: Int

	@ScaledMetric(relativeTo: .body) private var baseSize: CGFloat = 17

	private var fontSize: CGFloat { baseSize * CGFloat(scalePercent) / 100 }

	private var lineSpacing: CGFloat {
		switch lineSpacingChoice {
		case 1: return fontSize * 0.5
		case 2: return fontSize
		default: return 0
		}
	}

	private var paragraphPadding: CGFloat {
		switch paragraphSpacingChoice {
		case 1: return fontSize * 0.35
		case 2: return fontSize * 0.7
		default: return 2
		}
	}

	/// SwiftUI's Text has no justified alignment, so the desktop's fourth choice is not offered
	/// and a justified value carried in from elsewhere reads as leading.
	private var alignment: TextAlignment {
		switch alignmentChoice {
		case 1: return .center
		case 2: return .trailing
		default: return .leading
		}
	}

	private var frameAlignment: Alignment {
		switch alignmentChoice {
		case 1: return .center
		case 2: return .trailing
		default: return .leading
		}
	}

	func body(content: Content) -> some View {
		content
			.font(.system(size: fontSize))
			.lineSpacing(lineSpacing)
			.multilineTextAlignment(alignment)
			.padding(.vertical, paragraphPadding)
			.frame(maxWidth: .infinity, alignment: frameAlignment)
	}
}

extension View {
	func readabilityStyle(_ viewModel: AppViewModel) -> some View {
		modifier(
			ReadabilityStyle(
				scalePercent: viewModel.textScalePercent,
				lineSpacingChoice: viewModel.lineSpacingChoice,
				paragraphSpacingChoice: viewModel.paragraphSpacingChoice,
				alignmentChoice: viewModel.textAlignmentChoice
			)
		)
	}
}

File: android/app/src/main/kotlin/dev/paperback/android/ui/ReadabilityStyle.kt
package dev.paperback.android.ui

import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.Immutable
import androidx.compose.runtime.remember
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp

/**
 * The reader's readability preferences, resolved into the values a text view needs: a [TextStyle]
 * carrying size, line height and alignment, and the padding to put above and below each paragraph.
 */
@Immutable
data class ReadabilityStyle(
	val textStyle: TextStyle,
	val paragraphSpacing: Dp
)

/**
 * Builds the [ReadabilityStyle] for the reader's current preferences.
 *
 * The size is a multiplier on the body text style rather than a fixed point size. Compose measures
 * that style in sp, which already carries the system font scale, so a reader who has enlarged text
 * device-wide keeps that and scales it further here.
 */
@Composable
fun rememberReadabilityStyle(
	scalePercent: Int,
	lineSpacingChoice: Int,
	paragraphSpacingChoice: Int,
	alignmentChoice: Int
): ReadabilityStyle {
	val base = MaterialTheme.typography.bodyLarge
	val density = LocalDensity.current
	return remember(base, density, scalePercent, lineSpacingChoice, paragraphSpacingChoice, alignmentChoice) {
		val fontSize = base.fontSize * (scalePercent / 100f)
		// The body style's own line height is what "Normal" means, so the wider choices add to it
		// instead of replacing it. Dropping to a bare multiple of the font size would make Normal
		// tighter than the rest of the app.
		val extraLineSpacing = when (lineSpacingChoice) {
			1 -> 0.5f
			2 -> 1f
			else -> 0f
		}
		val paragraphSpacing = with(density) {
			when (paragraphSpacingChoice) {
				1 -> (fontSize * 0.35f).toDp()
				2 -> (fontSize * 0.7f).toDp()
				else -> 4.dp
			}
		}
		val alignment = when (alignmentChoice) {
			1 -> TextAlign.Center
			2 -> TextAlign.End
			3 -> TextAlign.Justify
			else -> TextAlign.Start
		}
		ReadabilityStyle(
			textStyle = base.copy(
				fontSize = fontSize,
				lineHeight = fontSize * (base.lineHeight.value / base.fontSize.value + extraLineSpacing),
				textAlign = alignment
			),
			paragraphSpacing = paragraphSpacing
		)
	}
}
